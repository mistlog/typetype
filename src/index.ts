export * from "./parser";
export * from "./generator";
export * from "./core";