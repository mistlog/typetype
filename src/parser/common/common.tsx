import { ReactPeg } from "react-peg";

// white space
export function _() {
    return (
        <repeat type="*">
            <set> \t\n\r</set>
        </repeat>
    );
}

export const Text = (content: string) => {
    return (
        <list>
            <_ /><text>{content}</text><_ />
        </list>
    )
}