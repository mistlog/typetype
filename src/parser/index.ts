export * from "./common";
export * from "./basic";
export * from "./function";
export * from "./expression";
export * from "./statement";
export * from "./declaration";
export * from "./start";