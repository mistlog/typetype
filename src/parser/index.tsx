export * from "./common";
export * from "./basic";
export * from "./expression";
export * from "./statement";
export * from "./declaration";
export * from "./start";