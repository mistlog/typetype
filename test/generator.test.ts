import { loadAST, saveCode, generateCode } from "./common";
import { TSTypeAliasDeclaration, TSType, TSTypeAliasDeclarationWithParams, TSFile, ITypeVariableDeclaration, ITypeFunctionDeclaration, ITypeFile } from "../src";

describe("type file", () => {
    test("TypeFile", () => {
        const ast = loadAST(`TypeFile.json`) as ITypeFile;
        const tsAST = TSFile(ast);
        expect(generateCode(tsAST)).toMatchSnapshot();
    })
})

describe("declaration", () => {
    test("TypeVariableDeclaration", () => {
        const ast = loadAST(`TypeVariableDeclaration.json`) as ITypeVariableDeclaration;
        const tsAST = TSTypeAliasDeclaration(ast);
        expect(generateCode(tsAST)).toMatchSnapshot();
    })

    test("TypeVariableDeclaration: context type", () => {
        const ast = loadAST(`TypeVariableDeclaration-ContextType.json`) as ITypeVariableDeclaration;
        const tsAST = TSTypeAliasDeclaration(ast);
        expect(generateCode(tsAST)).toMatchSnapshot();
    })

    test("TypeVariableDeclaration: conditional type", () => {
        const ast = loadAST(`TypeVariableDeclaration-ConditionalType.json`) as ITypeVariableDeclaration;
        const tsAST = TSTypeAliasDeclaration(ast);
        expect(generateCode(tsAST)).toMatchSnapshot();
    })

    test("TypeVariableDeclaration: export", () => {
        const ast = loadAST(`TypeVariableDeclaration-Export.json`) as ITypeVariableDeclaration;
        const tsAST = TSTypeAliasDeclaration(ast);
        expect(generateCode(tsAST)).toMatchSnapshot();
    })

    test("TypeVariableDeclaration: keyof", () => {
        const ast = loadAST(`TypeVariableDeclaration-KeyOf.json`) as ITypeVariableDeclaration;
        const tsAST = TSTypeAliasDeclaration(ast);
        expect(generateCode(tsAST)).toMatchSnapshot();
    })

    test("TypeVariableDeclaration: function", () => {
        const ast = loadAST(`TypeVariableDeclaration-Function.json`) as ITypeVariableDeclaration;
        const tsAST = TSTypeAliasDeclaration(ast);
        expect(generateCode(tsAST)).toMatchSnapshot();
    })

    test("TypeFunctionDeclaration", () => {
        const ast = loadAST(`TypeFunctionDeclaration.json`) as ITypeFunctionDeclaration;
        const tsAST = TSTypeAliasDeclarationWithParams(ast);
        expect(generateCode(tsAST)).toMatchSnapshot();
    })

    test("TypeFunctionDeclaration: default param", () => {
        const ast = loadAST(`TypeFunctionDeclaration-DefaultParam.json`) as ITypeFunctionDeclaration;
        const tsAST = TSTypeAliasDeclarationWithParams(ast);
        expect(generateCode(tsAST)).toMatchSnapshot();
    })

    test("TypeFunctionDeclaration: export", () => {
        const ast = loadAST(`TypeFunctionDeclaration-Export.json`) as ITypeFunctionDeclaration;
        const tsAST = TSTypeAliasDeclarationWithParams(ast);
        expect(generateCode(tsAST)).toMatchSnapshot();
    })
})


describe("ts-type", () => {
    function toCode(name: string) {
        const ast = loadAST(`${name}.json`);
        const tsAST = TSType(ast);
        const code = generateCode(tsAST);
        saveCode(code, `${name}`)
        return code;
    }
    test("ParenthesizedType", () => {
        const name = "ParenthesizedType";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("MappedType", () => {
        const name = "MappedType";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("MappedType: as", () => {
        const name = "MappedType-As";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("FunctionType", () => {
        const name = "FunctionType";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("FunctionType: constructor", () => {
        const name = "FunctionType-Constructor";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("FunctionType: optional param", () => {
        const name = "FunctionType-OptionalParam";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("FunctionType: params", () => {
        const name = "FunctionType-Params";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("FunctionType: rest", () => {
        const name = "FunctionType-Rest";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("FunctionType: nested", () => {
        const name = "FunctionType-Nested";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("FunctionType: type param", () => {
        const name = "FunctionType-TypeParam";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("ArrayType", () => {
        const name = "ArrayType";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("ArrayType: readonly", () => {
        const name = "ArrayType-Readonly";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("ArrayType: any", () => {
        const name = "ArrayType-Any";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("ArrayType: deep", () => {
        const name = "ArrayType-Deep";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("IndexType", () => {
        const name = "IndexType";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("IndexType: deep", () => {
        const name = "IndexType-Deep";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("StringTypeLiteral", () => {
        const name = "String-SpecialChar";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("UnionType", () => {
        const name = "UnionType";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("UnionType: special char", () => {
        const name = "UnionType-SpecialChar";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("IntersectionType", () => {
        const name = "IntersectionType";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("TypeArrowFunctionExpression", () => {
        const name = "TypeArrowFunctionExpression";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("TypeArrowFunctionExpression: with params", () => {
        const name = "TypeArrowFunctionExpression-WithParams";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("TypeArrowFunctionExpression: with infer", () => {
        const name = "TypeArrowFunctionExpression-Infer";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("TupleType: with infer", () => {
        const name = "TupleType-Infer";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("TupleType: rest", () => {
        const name = "TupleType-Rest";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("TupleType: readonly", () => {
        const name = "TupleType-Readonly";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("ObjectType", () => {
        const name = "ObjectType";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("ObjectType: whitespace in key", () => {
        const name = "ObjectType-Key-WhiteSpace";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("ObjectType: call signature", () => {
        const name = "ObjectType-CallSignature";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("ObjectType: shorthand", () => {
        const name = "ObjectType-Shorthand";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("ObjectType: spread", () => {
        const name = "ObjectType-Spread";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("ObjectType: modifier", () => {
        const name = "ObjectType-Modifier";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("ConditionalTypeExpression", () => {
        const name = "ConditionalTypeExpression";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("TemplateTypeLiteral", () => {
        const name = "TemplateTypeLiteral";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("ConditionalTypeExpression: 2", () => {
        const name = "ConditionalTypeExpression-2";
        expect(toCode(name)).toMatchSnapshot();
    })
})

describe("examples", () => {
    function toCode(name: string) {
        const ast = loadAST(`${name}.json`) as ITypeFunctionDeclaration;
        const tsAST = TSTypeAliasDeclarationWithParams(ast);
        const code = generateCode(tsAST);
        saveCode(code, `${name}`)
        return code;
    }

    test("readonly", () => {
        const name = "Example-readonly";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("pick", () => {
        const name = "Example-pick";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("parseURL", () => {
        const name = "Example-parseURL";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("parseURL: 2", () => {
        const name = "Example-parseURL-2";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("parseProtocol", () => {
        const name = "Example-parseProtocol";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("parseUserInfo", () => {
        const name = "Example-parseUserInfo";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("parseAuthority", () => {
        const name = "Example-parseAuthority";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("_isNumberString", () => {
        const name = "Example-_isNumberString";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("parsePort", () => {
        const name = "Example-parsePort";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("isNumberString", () => {
        const name = "Example-isNumberString";
        expect(toCode(name)).toMatchSnapshot();
    })

    test("parseHost", () => {
        const name = "Example-parseHost";
        expect(toCode(name)).toMatchSnapshot();
    })
})