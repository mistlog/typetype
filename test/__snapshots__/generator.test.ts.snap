// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`declaration TypeFunctionDeclaration 1`] = `"type TypeName<T> = T extends string ? \\"string\\" : \\"number\\";"`;

exports[`declaration TypeFunctionDeclaration: export 1`] = `"export type TypeName<T> = T extends string ? \\"string\\" : \\"number\\";"`;

exports[`declaration TypeVariableDeclaration 1`] = `"type result = string;"`;

exports[`declaration TypeVariableDeclaration: conditional type 1`] = `"type result = string extends string ? string : number;"`;

exports[`declaration TypeVariableDeclaration: export 1`] = `"export type result = string;"`;

exports[`declaration TypeVariableDeclaration: function 1`] = `"type result = (a: number, b: string) => void;"`;

exports[`declaration TypeVariableDeclaration: keyof 1`] = `
"type result = keyof {
  a: 1;
  b: 2;
};"
`;

exports[`examples _isNumberString 1`] = `"type _isNumberString<text> = text extends \\"\\" ? true : text extends \`\${infer digit}\${infer rest}\` ? digit extends Digit ? _isNumberString<rest> : false : false;"`;

exports[`examples isNumberString 1`] = `"type isNumberString<text> = text extends \\"\\" ? never : _isNumberString<text>;"`;

exports[`examples parseAuthority 1`] = `
"type parseAuthority<text> = text extends \`\${infer authority}@\${infer rest}\` ? {
  authority: parseUserInfo<authority>;
  rest: rest;
} : {
  authority: null;
  rest: text;
};"
`;

exports[`examples parseHost 1`] = `
"type parseHost<text> = text extends \`\${infer name}:\${infer port}\` ? parsePort<port> extends never ? never : {
  name: name;
  port: port;
} : {
  name: text;
};"
`;

exports[`examples parsePort 1`] = `"type parsePort<text> = isNumberString<text> extends true ? text : never;"`;

exports[`examples parseProtocol 1`] = `
"type parseProtocol<text> = text extends \`\${infer protocol}://\${infer rest}\` ? [{
  protocol: protocol;
}, rest] : never;"
`;

exports[`examples parseURL 1`] = `"type parseURL<text> = parseProtocol<text> extends [infer protocol, infer rest] ? object$assign<{}, [protocol, parseAuthority<rest>]> : never;"`;

exports[`examples parseURL: 2 1`] = `
"type parseURL<text> = parseProtocol<text> extends [infer protocol, infer rest] ? {
  protocol: protocol;
  rest: rest;
} : never;"
`;

exports[`examples parseUserInfo 1`] = `
"type parseUserInfo<text> = text extends \`\${infer username}:\${infer password}\` ? {
  username: username;
  password: password;
} : {
  username: text;
};"
`;

exports[`examples pick 1`] = `"export type pick<T, Keys extends keyof T> = { [K in Keys]: T[K] };"`;

exports[`examples readonly 1`] = `"export type MyReadonly<T> = { readonly [K in keyof T]: T[K] };"`;

exports[`ts-type ArrayType 1`] = `"string[]"`;

exports[`ts-type ArrayType: any 1`] = `"any[]"`;

exports[`ts-type ArrayType: deep 1`] = `"string[][]"`;

exports[`ts-type ConditionalTypeExpression 1`] = `"string extends string ? \\"string\\" : \\"number\\""`;

exports[`ts-type ConditionalTypeExpression: 2 1`] = `"T extends string ? \\"string\\" : T extends number ? \\"number\\" : never"`;

exports[`ts-type FunctionType 1`] = `"() => void"`;

exports[`ts-type FunctionType: nested 1`] = `"() => (a: number, b: string) => void"`;

exports[`ts-type FunctionType: params 1`] = `"(a: number, b: string) => number"`;

exports[`ts-type IndexType 1`] = `"obj[\\"a\\"]"`;

exports[`ts-type IndexType: deep 1`] = `"obj[\\"a\\"][2]"`;

exports[`ts-type IntersectionType 1`] = `"\\"0\\" & \\"1\\""`;

exports[`ts-type MappedType 1`] = `"{ [K in Keys]: boolean }"`;

exports[`ts-type MappedType: as 1`] = `"{ [K in Keys as \`get\${K}\`]: () => string }"`;

exports[`ts-type ObjectType 1`] = `
"{
  a: number;
  b: string;
  c: 1;
  d: \\"abc\\";
}"
`;

exports[`ts-type ObjectType: modifier 1`] = `
"{
  readonly a?: number;
}"
`;

exports[`ts-type ObjectType: shorthand 1`] = `
"{
  hello: hello;
}"
`;

exports[`ts-type ObjectType: spread 1`] = `
"object$assign<{}, [{
  hello: hello;
}, obj]>"
`;

exports[`ts-type TemplateTypeLiteral 1`] = `"\`hello \${World}\`"`;

exports[`ts-type TupleType: with infer 1`] = `"[infer protocol, infer rest]"`;

exports[`ts-type TypeArrowFunctionExpression 1`] = `"string extends string ? \\"string\\" : \\"number\\""`;

exports[`ts-type TypeArrowFunctionExpression: with infer 1`] = `"T extends infer U ? U : \\"number\\""`;

exports[`ts-type TypeArrowFunctionExpression: with params 1`] = `"T extends string ? \\"string\\" : \\"number\\""`;

exports[`ts-type UnionType 1`] = `"\\"0\\" | \\"1\\""`;

exports[`type file TypeFile 1`] = `
"type temp = string;
type func<T> = {
  a: T;
};"
`;
